AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt

```
@@ -713,8 +637,93 @@ class NoteEditor : AnkiActivity(), DeckSelectionListener, SubtitleListener, Tags
+                closeIntent = Intent().apply { putExtra(EXTRA_ID, context.intent.getStringExtra(EXTRA_ID)) }
+            } else if (!context.mEditFields!!.isEmpty()) {
+                context.mEditFields!!.first!!.focusWithKeyboard()
+            }
+
+            if (context.progressDialog != null && context.progressDialog!!.isShowing) {
+                try {
+                    context.progressDialog!!.dismiss()
+                } catch (e: IllegalArgumentException) {
+                    Timber.e(e, "Note Editor: Error on dismissing progress dialog")
+                }
+            }
+
+            if (closeEditorAfterSave) {
+                context.closeNoteEditor(closeIntent ?: Intent())
+            } else {
+                // Reset check for changes to fields
+                context.isFieldEdited = false
+                context.isTagsEdited = false
+            }
+        } else {
+            // RuntimeException occurred on adding note
+            context.closeNoteEditor(DeckPicker.RESULT_DB_ERROR, null)
+        }
+    }
+
+    /**
+     * Saves the [Note] to given [Collection]. Should be called from an IO thread/dispatcher
+     * @param col  Collection instance to which note is to be saved
+     * @param note Note instance to be saved
+     * @return Number of saved notes, null when any failure occurs
+     * */
+    private fun doInBackgroundAddNote(col: Collection, note: Note): Int? {
+        return try {
+            Timber.d("doInBackgroundAddNote - Started")
+            val noOfAddedCards = col.db.executeInTransaction {
+                col.addNote(note, Models.AllowEmpty.ONLY_CLOZE)
+            }
+            Timber.d("doInBackgroundAddNote - Cards Added $noOfAddedCards")
+            noOfAddedCards
+        } catch (e: RuntimeException) {
+            Timber.e(e, "doInBackgroundAddNote - RuntimeException on adding note")
+            CrashReportService.sendExceptionReport(e, "doInBackgroundAddNote")
+            null
+        }
+    }
```

I know this was discussed earlier, but having looked more closely at the code, I think it might be best to stop catching errors here, and return an Int instead of an Int?. The original code swallows the error, then calls context.closeNoteEditor(DeckPicker.RESULT_DB_ERROR, null) to close the editing screen. I think that's a bad idea - the user may have typed a lot of text in, and it gets discarded when an error occurs. Instead, I'd recommend you just rely on launchCatchingTask to display an error to the user. Once they close it, the stuff they typed in will still be there, and if they wish to preserve it, they could copy and paste it into a notes app, and manually close the screen when they're done. Without the try/catch, doInBackgroundAddNote() also becomes simpler, and it may be worth embedding the remaining code inside the withCol call below, and removing this separate function.

---

Yes, In fact first commit actually just a copy-paste. I've made the suggested changes in the last commit "Improved error handling while adding note"